Automatically emit `rustc-check-cfg` directives for AutoCfg

Techcable opened the pull request:
This avoids the `unexpected_cfgs` lint for any of the configs emitted by AutoCfg methods.
This lint is new in Rust 1.80: <https://blog.rust-lang.org/2024/05/06/check-cfg.html>

This adds a `emit_possibility` method to mirror the `emit` method.
Calling `emit` manually does not call `emit_possibility`.

Fixes issue #64

Techcable committed:
Automatically emit `rustc-check-cfg` directives for AutoCfg

This avoids the `unexpected_cfgs` lint for any of the configs emitted by AutoCfg methods.
This lint is new in Rust 1.80: <https://blog.rust-lang.org/2024/05/06/check-cfg.html>

This adds a `emit_possibility` method to mirror the `emit` method.
Calling `emit` manually does not call `emit_possibility`.

Fixes issue #64

cuviper commented on src/lib.rs:
```diff
@@ -98,6 +98,10 @@ pub struct AutoCfg {
 /// Cargo will use this in arguments to rustc, like `--cfg CFG`.
+///
+/// This does not automatically call [`emit_possibility`]
+/// so the compiler my generate an [`unexpected_cfgs` warning][check-cfg-flags].
```
This needs the `[check-cfg-flags]` link somewhere in the same doc block.

cuviper commented on src/lib.rs:
```diff
@@ -123,6 +127,25 @@ pub fn rerun_env(var: &str) {
 
+/// Indicates to rustc that a config flag should not generate an [`unexpected_cfgs` warning][check-cfg-flags]
+///
+/// This looks like `cargo:cargo:rustc-check-cfg=cfg(VAR)`
```
```suggestion
/// This looks like `cargo:rustc-check-cfg=cfg(VAR)`
```

cuviper commented on src/lib.rs:
```diff
@@ -357,13 +384,16 @@ impl AutoCfg {
     /// `_` in the generated config value.
     pub fn emit_has_path(&self, path: &str) {
+        let cfg_flag = format!("has_{}", mangle(path));
+        emit_possibility(&cfg_flag);
         if self.probe_path(path) {
-            emit(&format!("has_{}", mangle(path)));
+            emit(&cfg_flag);
         }
```
I think we can now simplify this to `self.emit_path_cfg(path, &format!(...));` -- It was being lazy about the formatting before, but now we can just forward the call. Same for the other `emit_has_` methods too.

cuviper reviewed (commented):
Thanks for working on this!

It would be nice to have a full test of this in CI, but it would need a real crate to get Cargo etc. to use it. If you're up for it, I'd create one or more nested crates under `ci/` that exercise this, probably denying `unexpected_cfgs` with both positive and negative cases. Then invoke those in the CI workflow, not recursive `cargo test` or anything. :)

Techcable committed:
Fix minor doc issues

Misisng link and incorrect `cargo:cargo:...` prefix

Techcable committed:
Have emit_has methods delegate wherever possible

Per request of @cuviper in PR #70

Techcable committed:
Add ci test crate for check-cfg

cuviper commented on ci/verify-check-cfg/Cargo.toml:
```diff
@@ -0,0 +1,12 @@
+name = "autocfg-verify-check-cfg"
+description = "A dummy crate to verify autocfg is emitting check-cfg directives"
+version = "0.1.0"
+edition = "2021"
```
I think this test could work even on ancient Rust (`edition = "2015"`). At a glance, it will just need to add `extern crate autocfg;` and change to `std::u32::MAX`.

Techcable replied:
Fixed with commit 027f73a911b748aecf70b0fc9f904fcf670f5879

Techcable committed:
Attempt rust 1.0 compat for ci/verify-check-cfg

I can't use rust 1.0 on my computer (M1 Mac),
so I am relying on CI to check this.

Techcable commented:
I have verified that commit 027f73a911b748aecf70b0fc9f904fcf670f5879 makes `ci/verify-check-cfg` work on rust 1.0

cuviper commented:
Curious, the CI failure looks like a compiler race condition that's unlikely to be caused by your change. I can't reproduce it locally though...

cuviper commented:
Oh, that was #65, but this is the first time I've seen it here in CI.

cuviper commented on ci/verify-check-cfg/Cargo.toml:
```diff
@@ -0,0 +1,12 @@
+edition = "2015"
+# only for testing
+publish = false
+
```
Build scripts used to need explicit mention:
```suggestion

build = "build.rs"

```

cuviper commented on ci/verify-check-cfg/src/main.rs:
```diff
@@ -0,0 +1,42 @@
+#![deny(unexpected_cfgs)]
```
```suggestion
#![allow(unknown_lints)]
#![deny(unexpected_cfgs)]
```

cuviper reviewed (commented):
Suggesting minor tweaks for older versions in that CI crate.

cuviper committed:
Apply suggestions from code review

Techcable commented:
Thank you! 😁
